circuit/algorithms/src/keccak/hash.rs

```diff
@@ -0,0 +1,454 @@
+
+        /* The first part of Algorithm 3, θ:
+         *
+         * for x = 0 to 4 do
+         *   C[x] = a[x, 0]
+         *   for y = 1 to 4 do
+         *     C[x] = C[x] ⊕ a[x, y]
+         *   end for
+         * end for
+         */
+        let mut c = Vec::with_capacity(WEIGHT);
+        for x in 0..MODULO {
+            c.push(&a[x] ^ &a[x + MODULO] ^ &a[x + (2 * MODULO)] ^ &a[x + (3 * MODULO)] ^ &a[x + (4 * MODULO)]);
+        }
+
+        /* The second part of Algorithm 3, θ:
+         *
+         * for x = 0 to 4 do
+         *   D[x] = C[x−1] ⊕ ROT(C[x+1],1)
+         *   for y = 0 to 4 do
+         *     A[x, y] = a[x, y] ⊕ D[x]
+         *   end for
+         * end for
+         */
+        let mut d = Vec::with_capacity(WEIGHT);
+        for x in 0..MODULO {
+            d.push(&c[(x + 4) % MODULO] ^ Self::rotate_left(&c[(x + 1) % MODULO], 63));
+        }
+        let mut a_1 = Vec::with_capacity(WEIGHT * WEIGHT);
+        for y in 0..MODULO {
+            for x in 0..MODULO {
+                a_1.push(&a[x + (y * MODULO)] ^ &d[x]);
+            }
+        }
+
+        /* Algorithm 4, π:
+         *
+         * for x = 0 to 4 do
+         *   for y = 0 to 4 do
+         *     (X, Y) = (y, (2*x + 3*y) mod 5)
+         *     A[X, Y] = a[x, y]
+         *   end for
+         * end for
+         *
+         * Algorithm 5, ρ:
+         *
+         * A[0, 0] = a[0, 0]
+         * (x, y) = (1, 0)
+         * for t = 0 to 23 do
+         *   A[x, y] = ROT(a[x, y], (t + 1)(t + 2)/2)
+         *   (x, y) = (y, (2*x + 3*y) mod 5)
+         * end for
+         */
+        let mut a_2 = a_1.clone();
+        for y in 0..MODULO {
+            for x in 0..MODULO {
+                // This step combines the π and ρ steps into one.
+                a_2[y + ((((2 * x) + (3 * y)) % MODULO) * MODULO)] =
+                    Self::rotate_left(&a_1[x + (y * MODULO)], rotl[x + (y * MODULO)]);
+            }
+        }
+
+        /* Algorithm 2, χ:
+         *
+         * for y = 0 to 4 do
+         *   for x = 0 to 4 do
+         *     A[x, y] = a[x, y] ⊕ ((¬a[x+1, y]) ∧ a[x+2, y])
+         *   end for
+         * end for
+         */
+        let mut a_3 = Vec::with_capacity(WEIGHT * WEIGHT);
+        for y in 0..MODULO {
+            for x in 0..MODULO {
+                let a = &a_2[x + (y * MODULO)];
+                let b = &a_2[((x + 1) % MODULO) + (y * MODULO)];
+                let c = &a_2[((x + 2) % MODULO) + (y * MODULO)];
+                a_3.push(a ^ ((!b) & c));
+            }
+        }
+
+        /* ι:
+         *
+         * A[0, 0] = A[0, 0] ⊕ RC
+         */
+        a_3[0] = &a_3[0] ^ round_constant;
+        a_3
+    }
+
+    /// Performs a rotate left operation on the given `u64` value.
+    fn rotate_left(value: &U64<E>, n: usize) -> U64<E> {
+        // Perform the rotation.
+        let bits_le = value.to_bits_le();
+        let bits_le = bits_le.iter().skip(n).chain(bits_le.iter()).take(64).cloned().collect::<Vec<_>>();
```

Comment:
```suggestion
        let mut bits_le = value.to_bits_le();
        bits_le.rotate_left(n);
```

Reply:
(ultimately it would be best to replace it with `Integer::rotate_left`, which could avoid the allocation of `to_bits_le`)